### PR TITLE
Rework README for 1.14

### PR DESCRIPTION
**Pull request by jsafrane**
- documentation for users should precede documentation of internal handling
- document all command line options
- document timeout handling
- use "external-attacher" everywhere

- [fix typo in cmdline argument description]

cc @kubernetes-csi/csi-misc, I'd like to update external-provisioner with very similar information, so let's polish it here.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,16 +2,82 @@
...
 # CSI attacher
 
-The csi-attacher is part of Kubernetes implementation of [Container Storage Interface (CSI)](https://github.com/container-storage-interface/spec).
+The external-attacher is a sidecar container that attaches volumes to nodes by calling `ControllerPublish` and `ControlerUnpublish` functions of CSI drivers. It is necessary because internal Attach/Detach controller running in Kubernetes controller-manager does not have any direct interfaces to CSI drivers.
+
+## Terminology
+
+In Kubernetes, term *attach* means 3rd party volume attachment to a node. This is common in cloud environments, where cloud API is able to attach a volume to a node without any code running on the node. In CSI terminology, this corresponds to `ControllerPublish` call.
+
+*Detach* then means reverse operation, 3rd party volume detachment from a node, `ControllerUnpublish` in CSI terminology.
```
Comment (msau42) [inline README.md]: "then means" => "is the"

File: README.md [hunk 2]
```diff
@@ -2,16 +2,82 @@
...
+* `Probe`: The external-attacher re-tries calling Probe until the driver reports it's ready. It re-tries also when it receives timeout from `Probe` call. The external-attacher has no limit of retries. It is expected that ReadinessProbe on the driver container will catch case when the driver takes too long time to get ready.
+* `GetPluginInfo`, `GetPluginCapabilitiesRequest`, `ControllerGetCapabilities`: The external-attacher expects that these calls are quick and does not retry them on any error, including timeout. Instead, it assumes that the driver is faulty and exits. Note that Kubernetes will likely start a new attacher container and it will start with `Probe` call.
+
+Correct timeout value depends on the storage backend and how quickly it is able to processes `ControllerPublish` and `ControllerUnpublish` calls. The value should be set to accommodate majority of them. It is fine if some calls time out - such calls will be re-tried after exponential backoff (starting with 5ms), however, this backoff will introduce delay when the call times out several times for a single volume.
+
+## Implementation details
```
Comment (msau42) [inline README.md]: The design details can probably go into a separate design document, and we can focus the main README on the usage

File: README.md [hunk 3]
```diff
@@ -2,16 +2,82 @@
...
+In Kubernetes, term *attach* means 3rd party volume attachment to a node. This is common in cloud environments, where cloud API is able to attach a volume to a node without any code running on the node. In CSI terminology, this corresponds to `ControllerPublish` call.
+
+*Detach* then means reverse operation, 3rd party volume detachment from a node, `ControllerUnpublish` in CSI terminology.
+
+It is **not** attach/detach operation performed by a code running on a node, such as attachment of iSCSI or Fibre Channel volumes. These are typically performed during `NodeStage` and `NodeUnstage` CSI calls and are not done by the external-attacher.
 
 ## Overview
+The external-attacher is an external controller that monitors `VolumeAttachment` objects created by controller-manager and attaches/detaches volumes to/from nodes (i.e. calls `ControllerPublish`/`ControllerUnpublish`. Full design can be found at Kubernetes proposal at [container-storage-interface.md](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md)
+
+## Usage
```
Comment (msau42) [inline README.md]: Add a compatibility matrix here, like how node-driver-registrar does

File: README.md [hunk 4]
```diff
@@ -2,16 +2,82 @@
...
+
+### Running on command line
```
Comment (msau42) [inline README.md]: Should we have a separate "development" file?

File: README.md [hunk 5]
```diff
@@ -2,16 +2,82 @@
...
+
+```sh
+kubectl create deploy/kubernetes/deployment.yaml
+```
+
+Note that the external-attacher does not scale with more replicas. Only one external-attacher is elected as leader and running. The others are waiting for the leader to die. They re-elect a new active leader in ~15 seconds after death of the old leader.
+
+### Command line options
+
+**Required arguments** that must be set:
```
Comment (msau42) [inline README.md]: Make these subheaders

File: README.md [hunk 6]
```diff
@@ -2,16 +2,82 @@
 
 # CSI attacher
 
-The csi-attacher is part of Kubernetes implementation of [Container Storage Interface (CSI)](https://github.com/container-storage-interface/spec).
+The external-attacher is a sidecar container that attaches volumes to nodes by calling `ControllerPublish` and `ControlerUnpublish` functions of CSI drivers. It is necessary because internal Attach/Detach controller running in Kubernetes controller-manager does not have any direct interfaces to CSI drivers.
+
+## Terminology
+
+In Kubernetes, term *attach* means 3rd party volume attachment to a node. This is common in cloud environments, where cloud API is able to attach a volume to a node without any code running on the node. In CSI terminology, this corresponds to `ControllerPublish` call.
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply README.md]: Fixed

File: README.md (hunk 2 above)
Comment (jsafrane) [reply README.md]: Fixed

File: README.md (hunk 4 above)
Comment (jsafrane) [reply README.md]: Moved to doc/development.md

File: README.md (hunk 5 above)
Comment (jsafrane) [reply README.md]: Fixed

File: README.md (hunk 3 above)
Comment (jsafrane) [reply README.md]: Added, although I am not sure I understand the matrix correctly.

File: README.md [hunk 7]
```diff
@@ -2,102 +2,78 @@
...
-
-## Design
-
-External attacher follows [controller](https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md) pattern and uses informers to watch for `VolumeAttachment` and `PersistentVolume` create/update/delete events. It filters out `VolumeAttachment` instances with `Attacher==<CSI driver name>` and processes these events in workqueues with exponential backoff. Real handling is deferred to `Handler` interface.
-
-`Handler` interface has two implementations, trivial and real one.
+## Terminology
 
-### Trivial handler
+In Kubernetes, the term *attach* means 3rd party volume attachment to a node. This is common in the cloud environments, where cloud API is able to attach a volume to a node without any code running on the node. In CSI terminology, this corresponds to the `ControllerPublish` call.
```
Comment (msau42) [inline README.md]: Sorry I was not clear earlier:
"the cloud environments" -> "cloud environments"
"cloud API" -> "the cloud API"

File: README.md [hunk 8]
```diff
@@ -2,102 +2,78 @@
...
 
-Trivial handler will be used for CSI drivers that don't support `ControllerPublish` calls and marks all `VolumeAttachment` as attached. It does not use any finalizers. This attacher can also be used for testing.
+*Detach* is the reverse operation, 3rd party volume detachment from a node, `ControllerUnpublish` in CSI terminology.
 
-### Real attacher
+It is **not** attach/detach operation performed by a code running on a node, such as attachment of iSCSI or Fibre Channel volumes. These are typically performed during `NodeStage` and `NodeUnstage` CSI calls and are not done by the external-attacher.
```
Comment (msau42) [inline README.md]: "not an attach/detach operation"
"such as an attachment"

File: README.md [hunk 9]
```diff
@@ -2,102 +2,78 @@
...
-  * Ignore PVs that don't have DeletionTimestamp.
-  * Checks that the PV is not used by any `VolumeAttachment` instance.
-  * Removes Attacher's finalizer on the PV if so.
-  * On any error, the PV is re-queued with exponential backoff.
+This information reflects the head of this branch.
 
-#### Concurrency
+| Compatible with CSI Version                                                                | Container Image                     | Min K8s Version |
+| ------------------------------------------------------------------------------------------ | ------------------------------------| --------------- |
+| [CSI Spec v0.3.0](https://github.com/container-storage-interface/spec/releases/tag/v0.3.0) | quay.io/k8scsi/csi-attacher:v0.2.0  | 1.10            |
```
Comment (msau42) [inline README.md]: I think we can remove the first two lines. We only need to have the matrix for the current HEAD of the branch, while the csi-docs can point to different branches.

File: README.md [hunk 10]
```diff
@@ -2,102 +2,78 @@
...
+| [CSI Spec v0.3.0](https://github.com/container-storage-interface/spec/releases/tag/v0.3.0) | quay.io/k8scsi/csi-attacher:v0.4.2  | 1.12            |
+| [CSI Spec v1.0.0](https://github.com/container-storage-interface/spec/releases/tag/v1.0.0) | quay.io/k8scsi/csi-attacher:v1.0.1  | 1.13            |
 
-Both PV queue and `VolumeAttachment` queue run in parallel. To ensure that removal of PV finalizers work without races:
+## Usage
 
-* The controller attaches PVs only when the PV has no DeletionTimestamp and has attacher's finalizer.
-* The controller removes finalizer only from PVs that have DeletionTimestamp.
+It is necessary to create a new service account and give it enough privileges to run the external-attacher. We provide one omnipotent yaml file that creates everything that's necessary, however it should be split into multiple files in production.
```
Comment (msau42) [inline README.md]: Looking at the yaml, it only contains attacher and the mock driver.  What could you split into more files?

(Also I think the Service is unnecessary)

File: README.md [hunk 11]
```diff
@@ -2,102 +2,78 @@
...
 
-As consequence, the attacher must be available until all PVs that refer to the CSI driver *are removed*. Even fully detached PVs have attacher's finalizer that is removed only after the PV is marked for deletion.
+```sh
+kubectl create deploy/kubernetes/deployment.yaml
+```
 
-#### Alternatives considered
+Note that the external-attacher does not scale with more replicas. Only one external-attacher is elected as leader and running. The others are waiting for the leader to die. They re-elect a new active leader in ~15 seconds after death of the old leader.
```
Comment (msau42) [inline README.md]: Is the 15 seconds configurable?

File: README.md [hunk 12]
```diff
@@ -2,102 +2,78 @@
...
 
-Secondary cache and locks was considered to keep a map PV -> list of VolumeAttachments that use the PV. Attacher's finalizer could be removed from a PV immediately after the last VolumeAttachment was deleted. Keeping this map is either racy or requires long critical sections with complicated error recovery.
+### Command line options
 
-## Usage
+#### Required arguments that must be set
+* `--csi-address <path to CSI socket>`: This is the path to the CSI driver socket inside the pod that the external-attacher container will use to issue CSI operations (e.g. `/csi/csi.sock`).
```
Comment (msau42) [inline README.md]: The code actually defaults this, although we don't use this default in any of our examples...

File: README.md (hunk 10 above)
Comment (jsafrane) [reply README.md]: Right, RBAC was cut out of there some time ago. I added a note about RBAC there.

File: README.md (hunk 11 above)
Comment (jsafrane) [reply README.md]: No. Do we need it configurable?

File: README.md (hunk 7 above)
Comment (jsafrane) [reply README.md]: Fixed

File: README.md (hunk 8 above)
Comment (jsafrane) [reply README.md]: fixed

File: README.md (hunk 9 above)
Comment (jsafrane) [reply README.md]: Fixe

File: README.md (hunk 12 above)
Comment (jsafrane) [reply README.md]: Added default and moved to "important options" below.

**Comment by jsafrane** [comment]
All comments resolved, PTAL

/hold
to remember to squash the patches before merge...

**Comment by msau42** [comment]
lgtm

**Comment by jsafrane** [comment]
Squashed.

/hold cancel

**Pull request merged**